Enhancement-02 PR 💯

boushrabettir opened the pull request:
**Please view the following changes below. Any code critics or suggestions would be *greatly appreciated*.**

- Moved CLI function to utility.py.
- Created blueprint_opportunity_formatter() for specific formatting.
- Created content_parser() to avoid DRY.
- Created merge_all_opportunity_data() to merge job and internship opportunities.
- Added an internship table.
- Added table name argument to some functions.
- Created 3 functions for internship opportunities from Github Repo, Apple, and LinkedIn.
- Added internship embed message in the payload.
- Updated .env.example.-
- Updated YAML file to `actions/checkout@v3`. Resolves [this issue!](https://github.com/acmcsufoss/acmOpportunities/issues/10)

--- 
**Here is an example response from the bot (using both jobs *and* internships):**

![image](https://github.com/acmcsufoss/acmOpportunities/assets/116927138/5a94527d-5678-4797-a83f-8745b3f84d0c)

boushrabettir committed:
Commenting code out in main.py to prevent workflow from triggering functions.

boushrabettir committed:
So *many* changes made! :hushed:

- Moved CLI function to `utility.py` as it was more fitting there and as I had to implement it in my `blueprint_opportunity_formatter()` function

- blueprint_opportunity_formatter() was created for *specific formatting* that currently only works with LinkedIn web scraping functions

- content_parser() was created because I noticed I kept not doing DRY (Don't Repeat Yourself). In other words, I kept repeating those same lines of code, so I ended up creating a function where it returns the parsed content.

- merge_all_opportunity_data() was created so it merges all the job opportunities in its respective list and then internships in their own respective list so when I utilize and call the GPT function, I don't need to call the function 3-4 times. Rather I can call it twice.

- Added an internship table for all internship opportunities

- In opportunity.py, some functions have a `table name` argument as we need to update the status, ingest opps, etc. into their respective tables.

- Created 3 different type of functions for internship opportunities sourcing from the following:
   1. Github Repo
   2. Apple
   3. LinkedIn

Am planning to add in more opportunities from bigger companies as well.

- Added in internship embed message in the payload

- Updated the .env.example with the following example links.

boushrabettir committed:
Updated YAML file to checkout v3.

boushrabettir commented:
Additionally, I am planning to complete enhancement #8 in this branch as well. 

Update: Completed.

boushrabettir committed:
Organized code by creating new file.

- Created new file called `gpt_utils.py` which holds all OpenAI functions which were originally in opportunity.py

- Moved `format_opportunities()` to opportunity.py as it fits better there.

- Organization fixes Issue #8

boushrabettir commented:
Isn't this cuter 😸?

boushrabettir commented:
Isn't this cuter? 🐱 

![image](https://github.com/acmcsufoss/acmOpportunities/assets/116927138/433d0223-a0f2-44b0-824a-c3c148590b5b)

boushrabettir commented:
I just realized for the [`Github Internship Repo`](https://github.com/pittcsc/Summer2024-Internships) am web scraping from, there can be multiple opportunities persay. 
Example:
![image](https://github.com/acmcsufoss/acmOpportunities/assets/116927138/0240b0d2-4f14-4288-8d24-11328e032d54)

Anyone have a suggestion for this?

From the top of my head, but is it possible to make the `_title` in the `Opportunity` class any type? Maybe placing `Any` as the type? There would be some slight modifications though here and there, but I am interested to see what everyone else has to say.

Currently, the code places both types of opportunity and the `main` link to the application page. People just need to search it up from there. I think that's a decent process, but again it could be optimized.

boushrabettir committed:
Converted from OPENAI :point_right: PaLm API

- Converted to PaLm API. There are a few issues with this where I am planning to figure out tomorrow.

- Fixed up embedded message aesthetic.

- env does not work in gpt_utils.py. Will fix this soon.

boushrabettir committed:
Merge branch 'main' into enhancement-02

boushrabettir commented:
There were some weird inconsistencies again, but everything seems to be working despite the `merge conflicts`.

![image](https://github.com/acmcsufoss/acmOpportunities/assets/116927138/3d32b261-c9ca-47ec-bd56-392619338216)

boushrabettir committed:
Fixed up function name as well remove OpenAI import.

Look at title.

boushrabettir commented:
For some reason, in `gpt_utils.py`, whenever I tried to do `from opportunity import Opportunity` I recieved this error.
```md
File "C:\Users\boush\OneDrive\Documents\GitHub\acmOpportunities\opportunity.py", line 4, in <module>
    import utility as utils
  File "C:\Users\boush\OneDrive\Documents\GitHub\acmOpportunities\utility.py", line 8, in <module>
    from opportunity import Opportunity
ImportError: cannot import name 'Opportunity' from partially initialized module 'opportunity' (most likely due to a circular import) (C:\Users\boush\OneDrive\Documents\GitHub\acmOpportunities\opportunity.py)
```

If anyone has any ideas, please let me know! I tried to figure it out, but I can't quite get to the reason why I was recieving this. I removed the import for now.

EthanThatOneKid commented on main.py:
```diff
@@ -234,36 +248,53 @@ async def execute_opportunities_webhook(webhook_url, message):
-async def main():
-    rapid_data = opps.gpt_job_analyzer(request_rapidapi_indeed_data())
-    linkedin_data = opps.gpt_job_analyzer(request_linkedin_data())
+# async def main():
+#     # job_opps = utils.merge_all_opportunity_data(
+#     #     request_rapidapi_indeed_data(), request_linkedin_data()
+#     # )
+#     # filtered_job_opps = gpt_utils.gpt_job_analyzer(job_opps)
+#     # opps.ingest_opportunities(filtered_job_opps, "jobs_table")
+#     # gpt_utils.gpt_job_analyzer(request_linkedin_data())
```
Nit: Take care of the unused commented code.

EthanThatOneKid replied:
I see this is intended to be used. I recommend making a new boolean flag for the script called called "dry-run" which when enabled, skips actually executing the operations.

EthanThatOneKid commented on main.py:
```diff
@@ -234,36 +248,53 @@ async def execute_opportunities_webhook(webhook_url, message):
-    rapid_data = opps.gpt_job_analyzer(request_rapidapi_indeed_data())
-    linkedin_data = opps.gpt_job_analyzer(request_linkedin_data())
+# async def main():
+#     # job_opps = utils.merge_all_opportunity_data(
```
Leave a comment explaining how this operation is optimal in that it minimizes the number of calls to the LLM server.

boushrabettir replied:
Updated [here](https://github.com/acmcsufoss/acmOpportunities/pull/11/commits/c8d33db4d61aec3478ddeec00a315509101800ef).

EthanThatOneKid commented on main.py:
```diff
@@ -234,36 +248,53 @@ async def execute_opportunities_webhook(webhook_url, message):
+#     #     request_rapidapi_indeed_data(), request_linkedin_data()
+#     # )
+#     # filtered_job_opps = gpt_utils.gpt_job_analyzer(job_opps)
+#     # opps.ingest_opportunities(filtered_job_opps, "jobs_table")
```
Additionally, leave a comment about how `filtered_job_opps` is for updating the "jobs_table" while the `filtered_internship_opps` is intended for the "internship_table" if we are to move forward with this database design.

boushrabettir replied:
Ended up adding new type column and changed table name to `opportunities_table`.

EthanThatOneKid commented on main.py:
```diff
@@ -200,7 +205,7 @@ def reset_processed_status():
 
 
-async def execute_opportunities_webhook(webhook_url, message):
+async def execute_opportunities_webhook(webhook_url, message, internship_message):
```
Here is a refactored example of the function that only populates the embed objects if there is a given message.

```py
async def execute_opportunities_webhook(webhook_url, message, internship_message):
    """
    Executes the message which receives the formatted message
    from the format_opportunities() function as well as the webhook
    URL for the respective Discord channel.
    """

    if not message and not internship_message:
        # No messages to send, so skip executing the webhook
        return

    # Create a dictionary payload for the message content
    payload = {
        "tts": False,
        "embeds": [
            {
                "title": f"✧･ﾟ: *✧･ﾟ:* 🎀 {date.today()} 🎀 ✧･ﾟ: *✧･ﾟ:*｡",
                "color": 0xFFFFFF,
            }
        ],
    }

    if message:
        # Add job opportunities embed if a message is provided
        payload["embeds"].append({
            "title": "¸„.-•~¹°”ˆ˜¨ JOB OPPORTUNITIES ¨˜ˆ”°¹~•-.„¸",
            "description": message,
            "color": 0x05A3FF,
        })

    if internship_message:
        # Add internship opportunities embed if a message is provided
        payload["embeds"].append({
            "title": " ¸„.-•~¹°”ˆ˜¨ INTERNSHIP OPPORTUNITIES ¨˜ˆ”°¹~•-.„¸",
            "description": internship_message,
            "color": 0x05A3FF,
        })

    # Convert the payload to JSON format
    json_payload = json.dumps(payload)

    # This will send a POST request to the webhook_url
    response = requests.post(
        webhook_url,
        data=json_payload,
        headers={"Content-Type": "application/json"}
    )

    if response.status_code == 204:
        print("Webhook message was sent successfully!")
    else:
        print(f"Failed to send webhook message. Status Code: {response.status_code}")
```

boushrabettir replied:
Implemented [here](https://github.com/acmcsufoss/acmOpportunities/pull/11/commits/fdeeb4acca854730a8315bcd1594017d233b4043).

EthanThatOneKid commented on utility.py:
```diff
@@ -1,6 +1,35 @@
 import os
+import argparse
+from bs4 import BeautifulSoup
+from opportunity import Opportunity
```
> For some reason, in `gpt_utils.py`, whenever I tried to do `from opportunity import Opportunity` I recieved this error.
> 
> ```lisp
> File "C:\Users\boush\OneDrive\Documents\GitHub\acmOpportunities\opportunity.py", line 4, in <module>
>     import utility as utils
>   File "C:\Users\boush\OneDrive\Documents\GitHub\acmOpportunities\utility.py", line 8, in <module>
>     from opportunity import Opportunity
> ImportError: cannot import name 'Opportunity' from partially initialized module 'opportunity' (most likely due to a circular import) (C:\Users\boush\OneDrive\Documents\GitHub\acmOpportunities\opportunity.py)
> ```
> 
> If anyone has any ideas, please let me know! I tried to figure it out, but I can't quite get to the reason why I was recieving this. I removed the import for now.

You may either import `opportunity.py` from `utility.py` or you may import `utility.py` from `opportunity.py`. But you may not do both.

boushrabettir replied:
Fixed [here](https://github.com/acmcsufoss/acmOpportunities/pull/11/commits/fdeeb4acca854730a8315bcd1594017d233b4043). 

Ended up deleting `gpt_utils.py` as it had no use and those functions can be easily integrated in `utility.py`.

EthanThatOneKid reviewed (commented):
Fantastic progress, @boushrabettir! I would recommend prioritizing my comment about the circular import error you encountered.

boushrabettir committed:
Some changes made.

- Deleted gpt_utils.py as I could integrate the same thing in utility.py. Placed all GPT functions in utility.py

- Fixed issue with links. Please view comments to understand implementation.

- Removed apple internship function since the Github Repo already has those opportunities. Also, it would cause a webscrape overload.

boushrabettir committed:
Added comment for the `merge_all_opportunity_data` function.

boushrabettir committed:
Updated comment again.

boushrabettir committed:
Updated .env.example accordingly to the .env file.

boushrabettir commented on opportunity.py:
```diff
@@ -1,23 +1,10 @@
-
 
 @dataclass
 class Opportunity:
```
**TO-DO**
- Change all the instances of the Opportunity class to without the underscore prefix
- Add a column to the `jobs_table` table with the column name of `type`
- Combine the two tables into one large table for all types of opportunities
- Every instance that the object is being constructed, do not forget to set the type of opportunity
@boushrabettir

boushrabettir replied:
Implemented [here](https://github.com/acmcsufoss/acmOpportunities/pull/11/commits/a0efeafe1f0d4b3cdcfac384dccd23bfeb895bba).

boushrabettir committed:
Started OpprtunityType of type Enum.

Review TO-DO to continue.

Work In Progress - https://github.com/acmcsufoss/acmOpportunities/pull/11#discussion_r1227597357

boushrabettir committed:
View Description.

After taking a few days off from the project, I am pleased to provide an update on the progress and ongoing maintenance of the project. The following enhancements and modifications have been implemented:

Enumerated OpportunityType: An OpportunityType enumeration has been incorporated into the project. This addition allows for a more structured and standardized representation of different types of opportunities.

New Enum Values: Two new values, CONFERENCE and SCHOLARSHIP, have been added to the OpportunityType enumeration. These values expand the range of opportunity types that can be defined within the project.

Opportunity Type Integration: The type of opportunity, based on the OpportunityType enumeration, has been integrated into the relevant functions in both opportunity.py and main.py. This ensures that the appropriate type information is captured and utilized throughout the project.

Column Name Updates: The column names have been adjusted to remove the underscore at the beginning. This update aligns with common naming conventions and improves the consistency and readability of the codebase.

Opportunity Type Column: A new column has been introduced to accommodate the storage of the opportunity type in the relevant data structures. This addition allows for efficient querying and retrieval of opportunities based on their type.

GitHub Scraping Function Enhancement: The GitHub scraping function has been updated to align with the latest requirements and specifications. This ensures accurate and reliable retrieval of data from GitHub as part of the project's functionality.

GPT Prompt Improvement: The GPT prompt has undergone minor revisions to address certain issues, such as responses being wrapped within an unnecessary double list structure. These adjustments enhance the quality and coherence of the AI model's generated responses.

jjoeldaniel commented on main.py:
```diff
@@ -37,43 +36,64 @@ def request_github_internship24_data() -> List[Opportunity]:
                 )
+                github_list.append(opportunity)
+
+            for i in t:
```
Not a docstring (line 74-82)

[Reference](https://peps.python.org/pep-0257/#what-is-a-docstring)

boushrabettir replied:
Implemented [here](https://github.com/acmcsufoss/acmOpportunities/pull/11/commits/1084fabdbf39ec53e031f7dd4998f397ca913f72)

EthanThatOneKid replied:
> Not a docstring (line 74-82)
> 
> [Reference](https://peps.python.org/pep-0257/#what-is-a-docstring)

Drag your mouse across the lines you intend to make your suggestion on.

![image](https://github.com/acmcsufoss/acmOpportunities/assets/31261035/d1b9e818-c266-4f99-ad92-3158b079eb9e)

jjoeldaniel commented on main.py:
```diff
@@ -37,43 +36,64 @@ def request_github_internship24_data() -> List[Opportunity]:
+                """
+                If the title length consists of one element,
+                this indicates that the link exists within
+                the company title and the company title only.
```
Not a docstring

[Reference](https://peps.python.org/pep-0257/#what-is-a-docstring)

boushrabettir replied:
Implemented [here](https://github.com/acmcsufoss/acmOpportunities/pull/11/commits/1084fabdbf39ec53e031f7dd4998f397ca913f72)

jjoeldaniel commented on main.py:
```diff
@@ -247,15 +276,15 @@ async def main():
     # )
     # filtered_internship_opps = utils.gpt_job_analyze(internship_opps)
-    # opps.ingest_opportunities(filtered_internship_opps, "internship_table")
+    # opps.ingest_opportunities(filtered_internship_opps)
```
Most of your `main()` seems to be commented out to avoid accidentally triggering the webhook on a dev branch. Can we instead change the workflow to [only trigger on main?](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#using-filters)

boushrabettir replied:
Implemented [here](https://github.com/acmcsufoss/acmOpportunities/pull/11/commits/1084fabdbf39ec53e031f7dd4998f397ca913f72)

boushrabettir committed:
Added specific branch for workflow. Fixed issue with comments being created as doc strings.

EthanThatOneKid reviewed (commented):
Thank you for reviewing this PR, @jjoeldaniel. In the future, please provide more context/reasoning in your suggestions for everyone's future reference. 

**Sample context**: Following the Google Python conventions[^1] of using multiline docstrings for documentation and hashtag comments for everything else ensures consistency within the project, making code easier to read and maintain.

[^1]: https://google.github.io/styleguide/pyguide.html#38-comments-and-docstrings

boushrabettir committed:
Merge branch 'main' into enhancement-02

boushrabettir committed:
Many changes made, please view description..

- Added prompts.py file which contains a dictionary with prompts for the respected type of opportunity

- Updated gpt_parsed_values() function to prevent any bugs that may occur when attempting to recieve a prompt from PaLM

- Created a limit for internship opportunities from Github to prevent char limitations in the embed

- Removed the conditional in the main function (lines 300-303) as no longer made sense in this case. Empty lists result in false as a boolean value

- Updated README file for more specification about how to properly use the cli command

- Updated recieve :arrow_right: receive

EthanThatOneKid commented on .github/workflows/automate.yaml:
Reminder: Update the environment variables of the "Run script" step to match `.env.example`.

boushrabettir replied:
Resolved [here](https://github.com/acmcsufoss/acmOpportunities/pull/11/commits/d557342dce7f8c21d782bf5d8342827199802401)

boushrabettir committed:
Updated yaml file and name for github internship key.

EthanThatOneKid reviewed (approved):
LGTM ![capycool](https://cdn.discordapp.com/emojis/1100136698475716618.webp?size=16&quality=lossless)

EthanThatOneKid commented on .github/workflows/automate.yaml:
```diff
@@ -3,14 +3,16 @@ name: Daily Data Retrieval Automation
 on:
   schedule:
     - cron: "0 21 */2 * *" # Runs every other day at 9PM UTC (2PM PST the same day)
+  branches:
+    - main
```
We do not want to trigger this workflow on push to main. This should not be here.

EthanThatOneKid replied:
Resolved in <https://github.com/acmcsufoss/acmOpportunities/commit/d777087dda4bd2fa7d258a3787ef9f7b757a6140>.
